src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_attend_and_excite.py

```diff
@@ -0,0 +1,964 @@
+from ...schedulers import KarrasDiffusionSchedulers
+from ...utils import logging
+from . import StableDiffusionPipeline, StableDiffusionPipelineOutput
+from .safety_checker import StableDiffusionSafetyChecker
+
+
+logger = logging.get_logger(__name__)
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import StableDiffusionAttendAndExcitePipeline
+
+        >>> pipe = StableDiffusionAttendAndExcitePipeline.from_pretrained(
+        ...     "CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16
+        ... ).to("cuda")
+
+        >>> pipe = pipe.to("cuda")
+
+        >>> prompt = "a cat and a frog"
+        >>> indices_to_alter = [2, 5]
+        >>> is_attend_and_excite = True
+        >>> seed = 6141
+        >>> g = torch.Generator("cuda").manual_seed(seed)
+
+        >>> images = pipe(
+        ...     prompt=[prompt],
+        ...     indices_to_alter=indices_to_alter,
+        ...     attention_res=16,
+        ...     guidance_scale=7.5,
+        ...     generator=g,
+        ...     num_inference_steps=50,
+        ...     max_iter_to_alter=25,
+        ...     run_standard_sd= not is_attend_and_excite,
+        ...     thresholds={0: 0.05, 10: 0.5, 20: 0.8},
+        ...     scale_factor=20,
+        ...     scale_range=(1.0, 0.5),
+        ...     smooth_attentions=True,
+        ...     sigma=0.5,
+        ...     kernel_size=3,
+        ...     ).images
+
+        >>> image = images[0]
+        >>> image.save(f"../images/{prompt}_{seed}_{ 'ae' if is_attend_and_excite else 'sd'}.png")
+
+        >>> image_explain = pipe.show_cross_attention(
+        ...    prompt=prompt,
+        ...    res=16, 
+        ...    from_where=("up", "down", "mid"),
+        ...    indices_to_alter=indices_to_alter,
+        ...    orig_image=image,
+        ...    )
+        >>> image_explain.save(f"../images/{prompt}_{seed}_{ 'ae' if is_attend_and_excite else 'sd'}_explain.png")
+        ```
+"""
+
+
+class AttentionControl(abc.ABC):
+    def step_callback(self, x_t):
+        return x_t
+
+    def between_steps(self):
+        return
+
+    @property
+    def num_uncond_att_layers(self):
+        return 0
+
+    @abc.abstractmethod
+    def forward(self, attn, is_cross: bool, place_in_unet: str):
+        raise NotImplementedError
+
+    def __call__(self, attn, is_cross: bool, place_in_unet: str):
+        if self.cur_att_layer >= self.num_uncond_att_layers:
+            self.forward(attn, is_cross, place_in_unet)
+        self.cur_att_layer += 1
+        if self.cur_att_layer == self.num_att_layers + self.num_uncond_att_layers:
+            self.cur_att_layer = 0
+            self.cur_step += 1
+            self.between_steps()
+
+    def reset(self):
+        self.cur_step = 0
+        self.cur_att_layer = 0
+
+    def __init__(self):
+        self.cur_step = 0
+        self.num_att_layers = -1
+        self.cur_att_layer = 0
+
+
+class AttentionStore(AttentionControl):
```

Comment:
Think `AttentionControl` is only used here - could we maybe fully remove it? We always try to reduce abstraction levels to a minimum

Reply:
yep! we can combine `AttentionStore` and `AttentionControl`